### DISCUSSION TITLE
Reminders not retro-actively triggered if device not unlocked after boot

### DISCUSSION THREAD

**Discussion by Mrnofish** [iSoron/uhabits#898] [Feature Requests]
## Pre-submission checklist
- [x] I am submitting a bug report, not a feature request.
- [x] I am running the latest version of Loop Habit Tracker.
- [x] I have have searched for similar issues, but did not find any matches.

## Description
Reminders are not retro-actively triggered if the device was on but not unlocked (encrypted device).

## Steps to reproduce
1. Set up a reminder for an early hour
2. Wake up, turn the device on, do not unlock, get busy with stuff. Leave it that way until after the reminder's trigger time has passed.
3. Now unlock the device.
4. No reminder for that habit will be triggered.

## System information
 - Phone: Oneplus 7 Pro
 - Phone Operating System: Android 10/Lineage 17.1
 - App version: 1.8.12 (F-Droid)

**Comment by hiqua** [comment]
To be clear, the device is on at the time when the reminder is supposed to be happen? I'm asking because you mention in your step 2. "turn the device on". If the device was off at this time, it's expected that the reminder would not happen. Your description seems to indicate that your device was on, just locked, which case is it?

**Comment by Mrnofish** [comment]
> Your description seems to indicate that your device was on, just locked, which case is it?

I think you are mistaking the notion of an encrypted device that is not yet unlocked after boot, with that of a device that is simply screen-locked.

> If the device was off at this time, it's expected that the reminder would not happen.

Be that as it may, I would still regard that as undesirable, and a defect: an application that is supposed to help users maintain their habits shouldn't "forget" to remind them just because the device was off at the time.

When it first starts, it should be checking for missed reminders and produce the appropriate notifications.

**Comment by hiqua** [comment]
Ok it's clearer now, thanks! So we're not talking about the lock screen, but a device which was off when the reminders were supposed to happen (the fact that it's encrypted does not a role by the way, or does it?).

I could double-check, but I do think it's intended, or at least it's just that retroactive reminders have never been implemented, not that they were thought to be there and working.

**Comment by Mrnofish** [comment]
> a device which was off when the reminders were supposed to happen

Nope, the device is on at the time the reminder is supposed to trigger, just not unlocked.

From the app's PoV, there may be little or no practical difference between the device being on and locked, and it being powered off entirely, what with /data being inaccessible.

> (the fact that it's encrypted does not a role by the way, or does it?).

It does, in that, in my understanding, an unencrypted device can access /data right after boot without having to wait for the user to unlock the encryption.

**Comment by iSoron** [comment]
This looks related to Android 7.0 Direct Boot mode. Skimming through the docs, it looks very hard to implement. I am not sure it's worth the effort:
- https://developer.android.com/training/articles/direct-boot
- https://www.androidauthority.com/direct-boot-711668-711668/

Detecting that a scheduled notification was not shown (e.g. because the device was off, locked, or because the O.S. killed the application) is more interesting to me, and could help with phones that have very aggressive power saving features.

**Comment by Mrnofish** [comment]
>  Skimming through the docs, it looks very hard to implement

I don't think I've seen an app yet, that is capable of displaying notifications before the encryption unlock. Perhaps I haven't noticed, but I simply assumed it wasn't possible.

> Detecting that a scheduled notification was not shown (e.g. because the device was off, locked, or because the O.S. killed the application) is more interesting to me

This would be great.
